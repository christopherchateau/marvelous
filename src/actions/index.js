export const storeCharacter = (character) => ({
  type: 'STORE_CHARACTER',
  character
})

export const updateStorageDetails = (currentIndex, length) => ({
  });
