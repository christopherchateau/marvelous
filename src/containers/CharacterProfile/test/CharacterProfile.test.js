import React from "react";
import { shallow } from "enzyme";
import { updateStorageDetails } from "../../../actions";
import {
  CharacterProfile,
  mapStateToProps,
  mapDispatchToProps
} from "../index";

describe("CharacterProfile", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <CharacterProfile
        dispatchStorageDetailsUpdate={jest.fn()}
        getCharacter={jest.fn()}
        storedCharacters={[
          { name: "Wolverine", id: 2, comics: [] },
          { name: "Spider-Man", id: 3, comics: [] },
          { name: "Hulk", id: 4, comics: [] }
        ]}
        characterCount={3}
        currentIndex={1}
        showFavorites={false}
      />
    );
  });

  it("should exist", () => {
    expect(wrapper).toBeDefined();
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render loading screen if less than 3 characters loaded", () => {
    wrapper = shallow(
      <CharacterProfile
        dispatchStorageDetailsUpdate={jest.fn()}
        getCharacter={jest.fn()}
        storedCharacters={[
          { name: "Wolverine", id: 2, comics: [] },
          { name: "Spider-Man", id: 3, comics: [] }
        ]}
        characterCount={3}
        currentIndex={1}
        showFavorites={false}
      />
    );
    expect(wrapper.find(".loading-screen")).toHaveLength(1);
  });

  it("should render character profile when 3+ characters loaded", () => {
    wrapper = shallow(
      <CharacterProfile
        dispatchStorageDetailsUpdate={jest.fn()}
        getCharacter={jest.fn()}
        storedCharacters={[
          { name: "Wolverine", id: 2, comics: [] },
          { name: "Spider-Man", id: 3, comics: [] },
          { name: "Hulk", id: 4, comics: [] }
        ]}
        characterCount={3}
        currentIndex={1}
        showFavorites={false}
      />
    );
    expect(wrapper.find(".CharacterProfile")).toHaveLength(1);
    expect(wrapper.find(".loading-screen")).toHaveLength(0);
  });

  it("should display no comics found message when zero covers loaded", () => {
    wrapper = shallow(
      <CharacterProfile
        dispatchStorageDetailsUpdate={jest.fn()}
        getCharacter={jest.fn()}
        storedCharacters={[
          { name: "Wolverine", id: 2, comics: [] },
          { name: "Spider-Man", id: 3, comics: [] },
          { name: "Hulk", id: 4, comics: [] }
        ]}
        characterCount={3}
        currentIndex={1}
        showFavorites={false}
      />
    );
    expect(wrapper.find(".no-comics-msg")).toHaveLength(1);
  });

  it("should display covers when url's provided", () => {
    wrapper = shallow(
      <CharacterProfile
        dispatchStorageDetailsUpdate={jest.fn()}
        getCharacter={jest.fn()}
        storedCharacters={[
          { name: "Wolverine", id: 2, comics: [] },
          { name: "Spider-Man", id: 3, comics: ["www.comic-cover.url"] },
          { name: "Hulk", id: 4, comics: [] }
        ]}
        characterCount={3}
        currentIndex={1}
        showFavorites={false}
      />
    );
    expect(wrapper.find(".comic-cover")).toHaveLength(1);
    expect(wrapper.find(".no-comics-msg")).toHaveLength(0);
  });

  describe("handleCardClick", () => {
    it("should call handleAarowClick with 'BACK' when left arrow icon clicked", () => {
      const spy = spyOn(wrapper.instance(), "handleArrowClick");
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().forceUpdate();

      wrapper.find(".fa-chevron-circle-left").simulate("click", mockEvent);
      expect(spy).toHaveBeenCalledWith("BACK");
    });

    it("should call handleAarowClick with 'FORWARD' when right arrow icon clicked", () => {
      const spy = spyOn(wrapper.instance(), "handleArrowClick");
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().forceUpdate();

      wrapper.find(".fa-chevron-circle-right").simulate("click", mockEvent);
      expect(spy).toHaveBeenCalledWith("FORWARD");
    });

    it("should dispatch storage details when handleArrowClick is run", () => {
      //const getCharacter = jest.fn()
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().forceUpdate();

      wrapper.find(".fa-chevron-circle-right").simulate("click", mockEvent);
      expect(wrapper.props().getCharacter).toHaveBeenCalled();
    });

    it.skip("should call getCharacter if index is beginning or end of array", () => {
    });
  });

  describe("updateIndex", () => {
    it("should increment index by one if direction is foward", () => {
      const result = wrapper.instance().updateIndex("FORWARD", 1);
      expect(result).toEqual(2);
    });

    it("should decrement index by one if direction is back", () => {
      const result = wrapper.instance().updateIndex("BACK", 1);
      expect(result).toEqual(0);
    });
  });

  describe("mapStateToProps", () => {
    let mockState;

    beforeEach(() => {
      mockState = {
        characters: [{ name: "Wolverine", id: 2 }],
        storageDetails: { currentIndex: 1, count: 3 }
      };
    });

    it("should return an array of characters", () => {
      const expected = [{ name: "Wolverine", id: 2 }];
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.storedCharacters).toEqual(expected);
    });

    it("should return count of stored characters", () => {
      const expected = 3;
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.characterCount).toEqual(expected);
    });

    it("should return current character index", () => {
      const expected = 1;
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps.currentIndex).toEqual(expected);
    });
  });

  describe("mapDispatchToProps", () => {
    it("should call dispatch storageDetailsUpdate when dispatchStorageDetailsUpdate is called", () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = updateStorageDetails({
        currentIndex: 1,
        count: 3
      });
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.dispatchStorageDetailsUpdate({
        currentIndex: 1,
        count: 3
      });
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
