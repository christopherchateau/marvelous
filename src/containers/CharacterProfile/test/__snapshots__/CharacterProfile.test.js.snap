// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterProfile should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterProfile
    characterCount={3}
    currentIndex={1}
    dispatchStorageDetailsUpdate={[MockFunction]}
    getCharacter={[MockFunction]}
    showFavorites={false}
    storedCharacters={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Spiderman loading"
          className="loading-pic"
          src="spiderman-loading.gif"
        />,
        <h3
          className="loading-msg"
        >
          Loading...
        </h3>,
      ],
      "className": "loading-screen",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "Spiderman loading",
          "className": "loading-pic",
          "src": "spiderman-loading.gif",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Loading...",
          "className": "loading-msg",
        },
        "ref": null,
        "rendered": "Loading...",
        "type": "h3",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Spiderman loading"
            className="loading-pic"
            src="spiderman-loading.gif"
          />,
          <h3
            className="loading-msg"
          >
            Loading...
          </h3>,
        ],
        "className": "loading-screen",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Spiderman loading",
            "className": "loading-pic",
            "src": "spiderman-loading.gif",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Loading...",
            "className": "loading-msg",
          },
          "ref": null,
          "rendered": "Loading...",
          "type": "h3",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
